File: backend/app/api/studio/emails.py
"""
Email Template endpoints - AI-generated HTML email templates.

Educational Note: Email templates demonstrate agent-based generation:
1. email_agent_executor orchestrates the generation
2. Claude creates HTML structure and content
3. Gemini generates header/banner images
4. Complete package: HTML + images

Agent Pattern:
- Uses email_agent_executor for orchestration
- Agent has tools for HTML generation and image creation
- Multi-step process with intermediate results
- Final output is a complete email template

Output Structure:
- HTML file with inline styles (email-safe)
- Image files for headers/banners
- All files stored in job-specific folder
- ZIP download available for full package

Routes:
- POST   /projects/<id>/studio/email-template              - Start generation
- GET    /projects/<id>/studio/email-jobs/<id>             - Job status
- GET    /projects/<id>/studio/email-jobs                  - List jobs
- GET    /projects/<id>/studio/email-templates/<file>      - Serve file
- GET    /projects/<id>/studio/email-templates/<id>/preview  - Preview HTML
- GET    /projects/<id>/studio/email-templates/<id>/download - Download ZIP
- DELETE /projects/<id>/studio/email-jobs/<id>             - Delete job
"""
import io
import re
import zipfile
from flask import g, jsonify, request, current_app, send_file, Response
from app.api.studio import studio_bp
from app.services.studio_services import studio_index_service
from app.services.tool_executors.email_agent_executor import email_agent_executor
from app.services.integrations.supabase import storage_service
from app.services.auth import require_permission


@studio_bp.route('/projects/<project_id>/studio/email-template', methods=['POST'])
@require_permission("studio", "emails")
def generate_email_template(project_id: str):
    """
    Start email template generation or edit via email agent.

    Request Body:
        - source_id: UUID of the source to generate template from (optional)
        - direction: User's direction/guidance (optional)
        - parent_job_id: UUID of the parent email job to edit (optional, for edits)
        - edit_instructions: Instructions for editing the parent email (optional, for edits)

    Response:
        - 202 Accepted with job_id for polling
    """
    try:
        data = request.get_json()

        # source_id is optional — email can be generated from direction alone
        source_id = data.get('source_id', '')
        direction = data.get('direction', '')

        # Edit mode: load parent job's HTML as context for refinement
        parent_job_id = data.get('parent_job_id')
        edit_instructions = data.get('edit_instructions')
        previous_markdown = None
        previous_title = None
        parent_source_name = None

        if parent_job_id:
            # Clean up any previously failed edit jobs for this parent
            try:
                all_jobs = studio_index_service.list_email_jobs(project_id)
                for job in all_jobs:
                    if (job.get("status") == "error"
                            and job.get("parent_job_id") == parent_job_id):
                        studio_index_service.delete_email_job(project_id, job["id"])
            except Exception:
                pass  # Non-critical cleanup — don't block the new edit

            parent_job = studio_index_service.get_email_job(project_id, parent_job_id)
            if not parent_job or not parent_job.get('html_file'):
                return jsonify({
                    'success': False,
                    'error': 'Parent job not found or has no content to edit'
                }), 404

            # Download previous HTML from Supabase Storage
            previous_markdown = storage_service.download_studio_file(
                project_id=project_id,
                job_type="emails",
                job_id=parent_job_id,
                filename=parent_job['html_file']
            )
            if previous_markdown is None:
                return jsonify({
                    'success': False,
                    'error': 'Failed to load parent email template content from storage'
                }), 500
            previous_title = parent_job.get('template_name')
            parent_source_name = parent_job.get('source_name')

        # Execute via email_agent_executor (creates job and launches agent)
        result = email_agent_executor.execute(
            project_id=project_id,
            source_id=source_id,
            direction=direction,
            user_id=g.user_id,
            edit_instructions=edit_instructions,
            previous_markdown=previous_markdown,
            previous_title=previous_title,
            parent_job_id=parent_job_id,
            parent_source_name=parent_source_name
        )

        if not result.get('success'):
            return jsonify(result), 400

        return jsonify({
            'success': True,
            'job_id': result['job_id'],
            'status': result['status'],
            'message': result['message']
        }), 202

    except Exception as e:
        current_app.logger.error(f"Error starting email template generation: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to start email template generation: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/email-jobs/<job_id>', methods=['GET'])
def get_email_job_status(project_id: str, job_id: str):
    """
    Get the status of an email template generation job.

    Response:
        - Job object with status, progress, and generated content
    """
    try:
        job = studio_index_service.get_email_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Email job {job_id} not found'
            }), 404

        return jsonify({
            'success': True,
            'job': job
        })

    except Exception as e:
        current_app.logger.error(f"Error getting email job status: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@studio_bp.route('/projects/<project_id>/studio/email-jobs', methods=['GET'])
def list_email_jobs(project_id: str):
    """
    List all email template jobs for a project.

    Query Parameters:
        - source_id: Optional filter by source

    Response:
        - List of email jobs (newest first)
    """
    try:
        source_id = request.args.get('source_id')
        jobs = studio_index_service.list_email_jobs(project_id, source_id)

        # Filter out orphaned failed-edit jobs (error + parent_job_id).
        # These are leftover from edit failures and should never be shown.
        # Deletion happens in the create flow to keep GET idempotent.
        clean_jobs = [
            job for job in jobs
            if not (job.get("status") == "error" and job.get("parent_job_id"))
        ]

        return jsonify({
            'success': True,
            'jobs': clean_jobs
        })

    except Exception as e:
        current_app.logger.error(f"Error listing email jobs: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@studio_bp.route('/projects/<project_id>/studio/email-templates/<filename>', methods=['GET'])
def get_email_template_file(project_id: str, filename: str):
    """
    Serve an email template file (HTML or image).

    Response:
        - HTML file or image file with appropriate headers
    """
    try:
        # Extract job_id from filename (format: {job_id}.html or {job_id}_image_N.ext or {job_id}_brand_logo.ext)
        # Job ID is the first UUID-length segment before any suffix
        job_id = filename.split('_image_')[0].split('_brand_')[0].split('.')[0]

        # Determine mimetype
        if filename.endswith('.html'):
            mimetype = 'text/html'
        elif filename.endswith('.png'):
            mimetype = 'image/png'
        elif filename.endswith('.jpg') or filename.endswith('.jpeg'):
            mimetype = 'image/jpeg'
        elif filename.endswith('.svg'):
            mimetype = 'image/svg+xml'
        else:
            mimetype = 'application/octet-stream'

        # Text files (HTML)
        if filename.endswith('.html'):
            content = storage_service.download_studio_file(
                project_id, "emails", job_id, filename
            )
            if content is None:
                return jsonify({
                    'success': False,
                    'error': f'File not found: {filename}'
                }), 404
            return Response(content, mimetype=mimetype, headers={'Content-Type': f'{mimetype}; charset=utf-8'})

        # Binary files (images)
        file_data = storage_service.download_studio_binary(
            project_id, "emails", job_id, filename
        )
        if file_data is None:
            return jsonify({
                'success': False,
                'error': f'File not found: {filename}'
            }), 404
        return send_file(io.BytesIO(file_data), mimetype=mimetype, as_attachment=False)

    except Exception as e:
        current_app.logger.error(f"Error serving email template file: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to serve file: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/email-templates/<job_id>/preview', methods=['GET'])
def preview_email_template(project_id: str, job_id: str):
    """
    Serve email template HTML for preview (iframe).

    Educational Note: The iframe can't send Authorization headers for <img> tags
    inside the HTML. We inject ?token= into image src URLs so the browser passes
    the JWT when fetching images. This rewriting only affects the preview — the
    original HTML on disk stays clean for download/export.

    Response:
        - HTML with auth-aware image URLs for rendering in iframe
    """
    try:
        # Get job to find HTML file
        job = studio_index_service.get_email_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Email job {job_id} not found'
            }), 404

        html_file = job.get('html_file')
        if not html_file:
            return jsonify({
                'success': False,
                'error': 'Email template not yet generated'
            }), 404

        # Download from Supabase Storage
        html_content = storage_service.download_studio_file(
            project_id, "emails", job_id, html_file
        )
        if html_content is None:
            return jsonify({
                'success': False,
                'error': f'HTML file not found: {html_file}'
            }), 404

        # Inject auth token into image URLs so <img> tags pass auth.
        # The token comes from the ?token= query param the frontend already sends.
        token = request.args.get('token', '')
        if token:
            # Match image src attributes pointing to our email-templates endpoint
            # e.g. src="/api/v1/projects/.../studio/email-templates/image.png"
            def _add_token(match: re.Match) -> str:
                url = match.group(1)
                separator = '&' if '?' in url else '?'
                return f'src="{url}{separator}token={token}"'

            html_content = re.sub(
                r'src="(/api/v1/projects/[^"]+/studio/email-templates/[^"]+\.(?:png|jpg|jpeg|gif|webp|svg))"',
                _add_token,
                html_content
            )

        return Response(html_content, mimetype='text/html')

    except Exception as e:
        current_app.logger.error(f"Error previewing email template: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to preview template: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/email-templates/<job_id>/download', methods=['GET'])
def download_email_template(project_id: str, job_id: str):
    """
    Download email template as ZIP file (HTML + images).

    Response:
        - ZIP file containing HTML and all images
    """
    try:
        # Get job to find files
        job = studio_index_service.get_email_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Email job {job_id} not found'
            }), 404

        html_file = job.get('html_file')
        if not html_file:
            return jsonify({
                'success': False,
                'error': 'Email template not yet generated'
            }), 404

        # Create ZIP in memory from Supabase Storage
        zip_buffer = io.BytesIO()
        with zipfile.ZipFile(zip_buffer, 'w', zipfile.ZIP_DEFLATED) as zip_file:
            # Add HTML file
            html_content = storage_service.download_studio_file(
                project_id, "emails", job_id, html_file
            )
            if html_content:
                zip_file.writestr(html_file, html_content)

            # Add image files
            images = job.get('images') or []
            for image_info in images:
                image_filename = image_info.get('filename')
                if image_filename:
                    image_data = storage_service.download_studio_binary(
                        project_id, "emails", job_id, image_filename
                    )
                    if image_data:
                        zip_file.writestr(image_filename, image_data)

        zip_buffer.seek(0)

        # Generate filename
        template_name = job.get('template_name', 'email_template')
        safe_name = "".join(c for c in template_name if c.isalnum() or c in (' ', '_', '-')).strip()
        zip_filename = f"{safe_name}.zip"

        return send_file(
            zip_buffer,
            mimetype='application/zip',
            as_attachment=True,
            download_name=zip_filename
        )

    except Exception as e:
        current_app.logger.error(f"Error downloading email template: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to download template: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/email-jobs/<job_id>', methods=['DELETE'])
def delete_email_job(project_id: str, job_id: str):
    """
    Delete an email template job and its files from Supabase Storage.

    Response:
        - Success status
    """
    try:
        job = studio_index_service.get_email_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Email job {job_id} not found'
            }), 404

        # Delete all files for this job from Supabase Storage
        storage_service.delete_studio_job_files(
            project_id=project_id,
            job_type="emails",
            job_id=job_id
        )

        # Delete job from index
        studio_index_service.delete_email_job(project_id, job_id)

        return jsonify({
            'success': True,
            'message': f'Email job {job_id} deleted'
        })

    except Exception as e:
        current_app.logger.error(f"Error deleting email job: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to delete email job: {str(e)}'
        }), 500

File: frontend/src/lib/api/studio/emails.ts
/**
 * Email Templates API
 * Educational Note: Handles AI-generated HTML email templates with images.
 * Uses an agentic approach for multi-step generation.
 */

import axios from 'axios';
import { API_BASE_URL } from '../client';
import type { JobStatus } from './index';
import { createLogger } from '@/lib/logger';

const log = createLogger('studio-emails-api');

/**
 * Email template section plan
 */
export interface EmailSection {
  section_type: 'header' | 'hero' | 'content' | 'product_grid' | 'cta' | 'testimonial' | 'footer';
  section_name: string;
  content_description: string;
  needs_image: boolean;
  image_description?: string;
}

/**
 * Email template color scheme
 */
export interface EmailColorScheme {
  primary: string;
  secondary: string;
  background: string;
  text: string;
  button: string;
}

/**
 * Generated email image info
 */
export interface EmailImage {
  section_name: string;
  filename: string;
  placeholder: string;
  url: string;
}

/**
 * Email template generation job
 */
export interface EmailJob {
  id: string;
  source_id: string;
  source_name: string;
  direction: string;
  status: JobStatus;
  status_message: string;
  error_message: string | null;
  // Edit lineage
  parent_job_id: string | null;
  edit_instructions: string | null;
  // Template plan
  template_name: string | null;
  template_type: 'newsletter' | 'promotional' | 'transactional' | 'announcement' | null;
  color_scheme: EmailColorScheme | null;
  sections: EmailSection[];
  layout_notes: string | null;
  // Generated content
  images: EmailImage[];
  html_file: string | null;
  html_url: string | null;
  preview_url: string | null;
  subject_line: string | null;
  preheader_text: string | null;
  // Metadata
  iterations: number | null;
  input_tokens: number | null;
  output_tokens: number | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/**
 * Response from starting email template generation
 */
export interface StartEmailResponse {
  success: boolean;
  job_id?: string;
  status?: string;
  message?: string;
  error?: string;
}

/**
 * Response from getting email job status
 */
export interface EmailJobStatusResponse {
  success: boolean;
  job?: EmailJob;
  error?: string;
}

/**
 * Response from listing email jobs
 */
export interface ListEmailJobsResponse {
  success: boolean;
  jobs: EmailJob[];
  error?: string;
}

/**
 * Email Templates API
 */
export const emailsAPI = {
  /**
   * Start email template generation or edit via email agent
   */
  async startGeneration(
    projectId: string,
    sourceId: string,
    direction?: string,
    parentJobId?: string,
    editInstructions?: string
  ): Promise<StartEmailResponse> {
    try {
      const body: Record<string, unknown> = {
        source_id: sourceId,
        direction: direction || '',
      };
      if (parentJobId) body.parent_job_id = parentJobId;
      if (editInstructions) body.edit_instructions = editInstructions;

      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/studio/email-template`,
        body
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to start email template generation');
      throw error;
    }
  },

  /**
   * Get the status of an email template job
   */
  async getJobStatus(projectId: string, jobId: string): Promise<EmailJobStatusResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/email-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to get email job status');
      throw error;
    }
  },

  /**
   * List all email template jobs for a project
   */
  async listJobs(projectId: string, sourceId?: string): Promise<ListEmailJobsResponse> {
    try {
      const params = sourceId ? { source_id: sourceId } : {};
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/email-jobs`,
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to list email jobs');
      throw error;
    }
  },

  /**
   * Get the full URL for an email template file (HTML or image)
   */
  getFileUrl(projectId: string, filename: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/email-templates/${filename}`;
  },

  /**
   * Get the preview URL for an email template
   */
  getPreviewUrl(projectId: string, jobId: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/email-templates/${jobId}/preview`;
  },

  /**
   * Get the download URL for an email template (ZIP)
   */
  getDownloadUrl(projectId: string, jobId: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/email-templates/${jobId}/download`;
  },

  /**
   * Delete an email template job
   */
  async deleteJob(projectId: string, jobId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/studio/email-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to delete email job');
      throw error;
    }
  },

  /**
   * Poll email job status until complete or error
   */
  async pollJobStatus(
    projectId: string,
    jobId: string,
    onProgress?: (job: EmailJob) => void,
    intervalMs: number = 2000,
    maxAttempts: number = 150  // Email generation can take longer (agentic)
  ): Promise<EmailJob> {
    let attempts = 0;
    let currentInterval = intervalMs;

    while (attempts < maxAttempts) {
      const response = await this.getJobStatus(projectId, jobId);

      if (!response.success || !response.job) {
        throw new Error(response.error || 'Failed to get job status');
      }

      const job = response.job;

      if (onProgress) {
        onProgress(job);
      }

      if (job.status === 'ready' || job.status === 'error') {
        return job;
      }

      await new Promise((resolve) => setTimeout(resolve, currentInterval));

      attempts++;

      if (attempts > 5 && currentInterval < 5000) {
        currentInterval = Math.min(currentInterval * 1.2, 5000);
      }
    }

    throw new Error('Email template generation timed out');
  },
};
